mcp/tests/staktrak/src/replay.ts

```diff
@@ -0,0 +1,1209 @@
+
+        await new Promise((resolve) => setTimeout(resolve, 300));
+        break;
+
+      case ActionType.INPUT:
+        await typeText(
+          element as HTMLInputElement,
+          action.value || "",
+          speedRef,
+          statusRef,
+          isTypingRef,
+          registerTimeout
+        );
+        break;
+
+      case ActionType.SELECT:
+        await selectOption(
+          element as HTMLSelectElement,
+          action.value || "",
+          speedRef,
+          statusRef,
+          registerTimeout
+        );
+        break;
+
+      case ActionType.CHECK:
+        (element as HTMLInputElement).checked = true;
+        element.dispatchEvent(new Event("change", { bubbles: true }));
+        break;
+
+      case ActionType.UNCHECK:
+        (element as HTMLInputElement).checked = false;
+        element.dispatchEvent(new Event("change", { bubbles: true }));
+        break;
+    }
+
+    if (statusRef.current !== ReplayStatus.PLAYING) return;
+
+    const nextAction = actionsRef.current[index + 1];
+    let delay = MIN_DELAY;
+
+    if (nextAction && action.timestamp && nextAction.timestamp) {
+      const timeDiff = nextAction.timestamp - action.timestamp;
+      delay =
+        Math.max(MIN_DELAY, Math.min(MAX_DELAY, timeDiff)) / speedRef.current;
+    }
+
+    timeoutRef.current = registerTimeout(
+      setTimeout(() => {
+        if (statusRef.current === ReplayStatus.PLAYING) {
+          setCurrentActionIndex(index + 1);
+          executeAction(
+            index + 1,
+            actionsRef,
+            statusRef,
+            cursorRef,
+            speedRef,
+            isTypingRef,
+            registerTimeout,
+            setCurrentActionIndex,
+            setStatus,
+            timeoutRef
+          );
+        } else {
+          console.warn(
+            `Not moving to next action because status is ${statusRef.current}`
+          );
+        }
+      }, delay)
+    );
+  } catch (error) {
+    console.error("Error executing action:", error);
+    window.parent.postMessage(
+      {
+        type: "staktrak-replay-error",
+        error: (error as Error).message,
+        action: action,
+      },
+      "*"
+    );
+
+    if (statusRef.current === ReplayStatus.PLAYING) {
+      timeoutRef.current = registerTimeout(
+        setTimeout(() => {
+          setCurrentActionIndex(index + 1);
+          executeAction(
+            index + 1,
+            actionsRef,
+            statusRef,
+            cursorRef,
+            speedRef,
+            isTypingRef,
+            registerTimeout,
+            setCurrentActionIndex,
+            setStatus,
+            timeoutRef
+          );
+        }, 2000)
+      );
+    }
+  }
+}
+
+export function initReplay(options?: ReplayOptions): void {
+  (window as any).convertToReplayActions = convertToReplayActions;
+
+  const replayStyles = createReplayStyles();
+  const cursor = createCursor();
+
+  (window as any).cursorRef = { current: cursor };

```

Comment:
why are all these attached to `window`? That is not necessary. Just make regular functions